### ISSUE TITLE
Guidance on registering a running broker using the CF CLI

### ISSUE THREAD

**Issue by joshisa** [couchbase/service-broker#70]
**Is your feature request related to a problem? Please describe.**
Thank you for this great project.  After reading through https://github.com/couchbase/service-broker/blob/master/documentation/modules/ROOT/pages/install/ingress.adoc, I'm struggling to imagine the best way to expose the service in a way that is compatible with the CF CLI.  The CF CLI create-service-broker expects 3 parameters:  user, password and url.  However, it's not obvious to me how to expose that after I have the couchbase broker service running in my cluster.

**Describe the solution you'd like**
I'd like to understand how to expose the running broker service in a way that supports the basic auth interaction that the CF CLI expects to interact with.

**Describe alternatives you've considered**
I'm using the KubeCF project and I hoped that maybe the internal DNS resolution might pick up on this broker service.  But the CF CLI acts as an intermediary and so the internal DNS visibility isn't useful here.

**Additional context**
I really appreciate the project.  I like the idea of having a generic K8s template that can be triggered with an OS API broker interaction.  Very useful.

**Comment by joshisa** [comment]
Hi @toabi - I noticed you've been able to test and register this with a CF marketplace in other posted issues.  I was wondering if you could give me any hints/tips to expose the broker service and use the CF CLI to `create-service-broker` to it.  I've tried using an ingress ... but I cannot figure out what would be the creds to talk with it via the CF CLI.  Thanks.

**Comment by spjmurray** [comment]
If the ask is a simple as basic password authentication I can have that released tomorrow. I was going to reserve that for RBAC, but seeing as CF don't support tokens (part of their own spec no less), then it's a required feature.

**Comment by toabi** [comment]
Old OSB Spec only supported Basic Auth so CF only supports that. I solved with traefik middlewares which add basic auth to the ingress and then add the http header with the static token.

**Comment by joshisa** [comment]
@spjmurray -  Inclusion of support for basic password auth would be fantastic.  Would make the integration much smoother and improve interoperability.  Thanks.

**Comment by spjmurray** [comment]
Sigh, that's how up to date Kubernetes' Service Catalog is... I'll add support for username/password keys in a secret, then you can either mount one, or have Hashicorp Vault inject one for you, then build in password/token/TLS rotation

**Comment by joshisa** [comment]
Thank you @spjmurray for such a quick response and merge.  I tested it out and it worked perfectly in terms of registering the broker with my `cf create-service-broker <user> <pw> <url>` flow.  I'm happy to close this issue.

One observation:  
I noticed that when I spun up a couchbase cluster via my `cf create-service couchbase-developer couchbase-developer-private couchbase-test -c '{"password": "mypassword"}'`, the Couchbase cluster pods spin up fine, **but the CF service status stays in an indefinite state of "create in progress"**.  Unfortunately, this makes it unbindable and difficult to delete (e.g. purge + k8s CouchbaseCluster delete).  

>Should I open a separate issue to track this or did I miss something in my configuration deployment for the broker?  

Thanks again.

```
▶ cf service couchbase-test
Showing info of service couchbase-test in org myorg / space myspace as admin...

name:             couchbase-test
service:          couchbase-developer
tags:
plan:             couchbase-developer-private
description:      Couchbase Server developer deployments
documentation:
dashboard:        https://instance-xkf820k0.default:18091
service broker:   couchbase

Showing status of last operation from service couchbase-test...

status:    create in progress
message:   resource couchbase.com/v2/CouchbaseCluster instance-xkf820k0 doesn't contain the condition Available
started:   2021-03-16T05:23:53Z
updated:   2021-03-16T13:27:28Z

There are no bound apps for this service.

Upgrades are not supported by this broker.
```

**Comment by spjmurray** [comment]
Ah I see, looks like the cluster isn't coming up for some reason.  The operator logs probably have an error level message in there somewhere that's pertinent.

**Comment by joshisa** [comment]
I'll take a look into the logs and update this comment with what I find a bit later today.  
@spjmurray - 
UPDATE:
Here's what I see repeatedly in the operator logs.  This line in particular catches my eye: `unexpected cluster phase: Failed`

```
{
  "level": "error",
  "ts": 1615912531.151255,
  "logger": "cluster",
  "msg": "Cluster setup failed",
  "cluster": "default/instance-xkf820k0",
  "error": "unexpected cluster phase: Failed",
  "stacktrace": "github.com/couchbase/couchbase-operator/vendor/github.com/go-logr/zapr.(*zapLogger).Error\n\t/home/couchbase/jenkins/workspace/couchbase-operator-build/goproj/src/github.com/couchbase/couchbase-operator/vendor/github.com/go-logr/zapr/zapr.go:128\ngithub.com/couchbase/couchbase-operator/pkg/cluster.New\n\t/home/couchbase/jenkins/workspace/couchbase-operator-build/goproj/src/github.com/couchbase/couchbase-operator/pkg/cluster/cluster.go:143\ngithub.com/couchbase/couchbase-operator/pkg/controller.(*CouchbaseClusterReconciler).Reconcile\n\t/home/couchbase/jenkins/workspace/couchbase-operator-build/goproj/src/github.com/couchbase/couchbase-operator/pkg/controller/controller.go:71\ngithub.com/couchbase/couchbase-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/home/couchbase/jenkins/workspace/couchbase-operator-build/goproj/src/github.com/couchbase/couchbase-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:215\ngithub.com/couchbase/couchbase-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1\n\t/home/couchbase/jenkins/workspace/couchbase-operator-build/goproj/src/github.com/couchbase/couchbase-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158\ngithub.com/couchbase/couchbase-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/home/couchbase/jenkins/workspace/couchbase-operator-build/goproj/src/github.com/couchbase/couchbase-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133\ngithub.com/couchbase/couchbase-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/home/couchbase/jenkins/workspace/couchbase-operator-build/goproj/src/github.com/couchbase/couchbase-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134\ngithub.com/couchbase/couchbase-operator/vendor/k8s.io/apimachinery/pkg/util/wait.Until\n\t/home/couchbase/jenkins/workspace/couchbase-operator-build/goproj/src/github.com/couchbase/couchbase-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88"
}{
  "level": "error",
  "ts": 1615912531.1824574,
  "logger": "controller",
  "msg": "Failed to create Couchbase cluster",
  "cluster": {
    "namespace": "default",
    "name": "instance-xkf820k0"
  },
  "error": "unexpected cluster phase: Failed",
  "stacktrace": "github.com/couchbase/couchbase-operator/vendor/github.com/go-logr/zapr.(*zapLogger).Error\n\t/home/couchbase/jenkins/workspace/couchbase-operator-build/goproj/src/github.com/couchbase/couchbase-operator/vendor/github.com/go-logr/zapr/zapr.go:128\ngithub.com/couchbase/couchbase-operator/pkg/controller.(*CouchbaseClusterReconciler).Reconcile\n\t/home/couchbase/jenkins/workspace/couchbase-operator-build/goproj/src/github.com/couchbase/couchbase-operator/pkg/controller/controller.go:73\ngithub.com/couchbase/couchbase-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/home/couchbase/jenkins/workspace/couchbase-operator-build/goproj/src/github.com/couchbase/couchbase-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:215\ngithub.com/couchbase/couchbase-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1\n\t/home/couchbase/jenkins/workspace/couchbase-operator-build/goproj/src/github.com/couchbase/couchbase-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158\ngithub.com/couchbase/couchbase-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/home/couchbase/jenkins/workspace/couchbase-operator-build/goproj/src/github.com/couchbase/couchbase-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133\ngithub.com/couchbase/couchbase-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/home/couchbase/jenkins/workspace/couchbase-operator-build/goproj/src/github.com/couchbase/couchbase-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134\ngithub.com/couchbase/couchbase-operator/vendor/k8s.io/apimachinery/pkg/util/wait.Until\n\t/home/couchbase/jenkins/workspace/couchbase-operator-build/goproj/src/github.com/couchbase/couchbase-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88"
}{
  "level": "error",
  "ts": 1615912531.1825776,
  "logger": "kubebuilder.controller",
  "msg": "Reconciler error",
  "controller": "couchbase-controller",
  "request": "default/instance-xkf820k0",
  "error": "unexpected cluster phase: Failed",
  "stacktrace": "github.com/couchbase/couchbase-operator/vendor/github.com/go-logr/zapr.(*zapLogger).Error\n\t/home/couchbase/jenkins/workspace/couchbase-operator-build/goproj/src/github.com/couchbase/couchbase-operator/vendor/github.com/go-logr/zapr/zapr.go:128\ngithub.com/couchbase/couchbase-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/home/couchbase/jenkins/workspace/couchbase-operator-build/goproj/src/github.com/couchbase/couchbase-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:217\ngithub.com/couchbase/couchbase-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1\n\t/home/couchbase/jenkins/workspace/couchbase-operator-build/goproj/src/github.com/couchbase/couchbase-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158\ngithub.com/couchbase/couchbase-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/home/couchbase/jenkins/workspace/couchbase-operator-build/goproj/src/github.com/couchbase/couchbase-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133\ngithub.com/couchbase/couchbase-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/home/couchbase/jenkins/workspace/couchbase-operator-build/goproj/src/github.com/couchbase/couchbase-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134\ngithub.com/couchbase/couchbase-operator/vendor/k8s.io/apimachinery/pkg/util/wait.Until\n\t/home/couchbase/jenkins/workspace/couchbase-operator-build/goproj/src/github.com/couchbase/couchbase-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88"
}
```

However, everything appears to be running from what I can tell.
![image](https://user-images.githubusercontent.com/1133849/111348838-f7f0e080-8656-11eb-8c63-e4b7ca76480f.png)

### UPDATE 2
```
kubectl logs instance-xkf820k0-0000
Starting Couchbase Server -- Web UI available at http://<ip>:8091 and logs available in /opt/couchbase/var/lib/couchbase/logs

leads me to ...

kubectl exec -it instance-xkf820k0-0000 -- cat /opt/couchbase/var/lib/couchbase/logs/error.log
[ns_server:error,2021-03-16T05:26:52.732Z,ns_1@instance-xkf820k0-0000.instance-xkf820k0.default.svc:service_status_keeper_worker<0.517.0>:rest_utils:get_json:62]Request to (indexer) getIndexStatus failed: {ok,
                                             {{500,"Internal Server Error"},
                                              [{"Content-Length","220"},
                                               {"Date",
                                                "Tue, 16 Mar 2021 05:26:52 GMT"},
                                               {"Content-Type",
                                                "application/json"}],
                                              <<"{\"code\":\"error\",\"error\":\"Fail to retrieve cluster-wide metadata from index service\",\"failedNodes\":[\"instance-xkf820k0-0001.instance-xkf820k0.default.svc:8091\",\"instance-xkf820k0-0002.instance-xkf820k0.default.svc:8091\"]}">>}}
[ns_server:error,2021-03-16T05:26:52.732Z,ns_1@instance-xkf820k0-0000.instance-xkf820k0.default.svc:service_status_keeper-index<0.518.0>:service_status_keeper:handle_cast:119]Service service_index returned incorrect status
[ns_server:error,2021-03-16T14:44:03.571Z,ns_1@instance-xkf820k0-0000.instance-xkf820k0.default.svc:<0.7244.37>:menelaus_web_alerts_srv:can_listen:528]Cannot listen due to nxdomain from inet:getaddr

[ns_server:error,2021-03-16T14:44:24.571Z,ns_1@instance-xkf820k0-0000.instance-xkf820k0.default.svc:<0.7999.37>:menelaus_web_alerts_srv:can_listen:528]Cannot listen due to nxdomain from inet:getaddr

[ns_server:error,2021-03-16T14:44:33.571Z,ns_1@instance-xkf820k0-0000.instance-xkf820k0.default.svc:<0.8342.37>:menelaus_web_alerts_srv:can_listen:528]Cannot listen due to nxdomain from inet:getaddr

[ns_server:error,2021-03-16T14:44:45.571Z,ns_1@instance-xkf820k0-0000.instance-xkf820k0.default.svc:<0.8782.37>:menelaus_web_alerts_srv:can_listen:528]Cannot listen due to nxdomain from inet:getaddr

[ns_server:error,2021-03-16T14:44:57.571Z,ns_1@instance-xkf820k0-0000.instance-xkf820k0.default.svc:<0.9211.37>:menelaus_web_alerts_srv:can_listen:528]Cannot listen due to nxdomain from inet:getaddr

[ns_server:error,2021-03-16T15:49:42.571Z,ns_1@instance-xkf820k0-0000.instance-xkf820k0.default.svc:<0.19449.41>:menelaus_web_alerts_srv:can_listen:528]Cannot listen due to nxdomain from inet:getaddr

[ns_server:error,2021-03-16T15:49:54.570Z,ns_1@instance-xkf820k0-0000.instance-xkf820k0.default.svc:<0.19866.41>:menelaus_web_alerts_srv:can_listen:528]Cannot listen due to nxdomain from inet:getaddr

[ns_server:error,2021-03-16T15:50:12.571Z,ns_1@instance-xkf820k0-0000.instance-xkf820k0.default.svc:<0.20539.41>:menelaus_web_alerts_srv:can_listen:528]Cannot listen due to nxdomain from inet:getaddr
```

### Update 3

It appears as though the above errors were older, early on alerts.  Looking at the dashboard view shows everything green.  The STATUS field for the CouchbaseCluster remains empty which I think is causing CF to see the service as unavailable.  

![image](https://user-images.githubusercontent.com/1133849/111372114-989fca00-8670-11eb-8ae5-9e1d55a479a7.png)

```
▶ kubectl get CouchbaseCluster
NAME                VERSION             SIZE   STATUS   UUID                               AGE
instance-xkf820k0   couchbase-6.5.1-1   1               1f21b71xxxxxxxxxxxxxxxxxxx3c709   15h
```

**Comment by joshisa** [comment]
Comparing status objects from doc and runtime ... My Condition Type is `Balanced` rather than `Available`

#### [Documentation example](https://docs.couchbase.com/operator/current/reference-couchbasecluster-status.html):
```
status:
  phase: Running
  reason: ""
  controlPaused: false
  conditions:
  - type: Available
    status: "True"
    reason: Available
    lastUpdateTime: "2020-01-29T11:46:22Z"
    lastTransitionTime: "2020-01-29T11:46:22Z"
```

#### My Cluster
```
status:
    [...]
    conditions:
    - lastTransitionTime: "2021-03-16T05:40:45Z"
      lastUpdateTime: "2021-03-16T05:40:45Z"
      message: Data is equally distributed across all nodes in the cluster
      reason: Balanced
      status: "True"
      type: Balanced
    currentVersion: couchbase-6.5.1-1
    members:
      ready:
      - instance-xkf820k0-0000
      - instance-xkf820k0-0001
      - instance-xkf820k0-0002
    size: 1
```

**Comment by spjmurray** [comment]
Ugh, bloody Cluster Failed 😸

So what should happen is the cluster comes up, the topology is reconciled, everything else is reconciled and then the Available condition is added. What I can assume is happened is that something went wrong between topology and available, the operator was restarted, and then it didn't like what it found on restart.

I'll give you a full set of debugging steps in the morning.

**Comment by spjmurray** [comment]
Suffice to say what I'd do (and I'm technical lead on the project :smile: ), is do a `kubectl logs -f deployment/couchbase-operator | tee logs` as soon as you make the provision call, that way you get logs from the start right up until bad happens.  Unless you are wise and already have fluentd collecting them for you.

**Comment by joshisa** [comment]
@spjmurray I love split screen keyboard orchestration 😄 and I'm not very wise.  With your advice, I've caught the runtime panic triggering the operator restart.  Log attached.

...
[logs.txt](https://github.com/couchbase/service-broker/files/6157766/logs.txt)

**Comment by spjmurray** [comment]
Oh, I'm just guessing here but did you install the dynamic admission crontoller too? That should be the only reason things aren't filled in like we expect.

**Comment by joshisa** [comment]
@spjmurray Thank you for all of your help.  You called it right.  I didn't have the DAC running. After getting that installed, I did have to make one small tweak in the example configuration for GeneratingCertificate (https://github.com/couchbase/service-broker/blob/master/examples/configurations/couchbase-server/broker.yaml#L443).  It wasn't liking my `cluster.local` domain extensions so I added a couple of wildcards for *-srv and * that also appended the `cluster.local` .  Otherwise, the certificate was being rejected.  I now have things up and running 🆒 .  I'm good with closing this issue.
